crates/amaru-minicbor-extra/src/decode/lazy.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2025 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use minicbor as cbor;
+use std::{fs, io::Read};
+
+/// A decoder that only consumes bytes CHUNK_SIZE at a time. Useful to decode large files while
+/// maintaining memory usage low.
+///
+/// The decoder keeps an internal state with the bytes that have been read but not consumed, and a
+/// handle to a source that implements [`std::io::Read`].
+///
+/// See [`Self::from_file`] for example, to lazily read and decode from a (large) file.
+pub struct LazyDecoder<'a> {
+    reader: &'a mut dyn Read,
+    bytes: Vec<u8>,
+}
+
+impl<'a> LazyDecoder<'a> {
+    const CHUNK_SIZE: usize = 2 * 1024 * 1024; // 2MiB, chosen at random by fair dice roll
+
+    pub fn from_file(file: &'a mut fs::File) -> Self {
+        Self {
+            reader: file,
+            bytes: Vec::with_capacity(Self::CHUNK_SIZE),
+        }
+    }
+
+    /// Consumes enough bytes and skip the next CBOR element.
+    pub fn skip(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        self.with_decoder(|d| Ok(d.skip()?))
+    }
+
+    /// Consumes enough bytes and decode the next CBOR element.
+    pub fn decode<T: for<'d> cbor::decode::Decode<'d, ()>>(
+        &mut self,
+    ) -> Result<T, Box<dyn std::error::Error>> {
+        self.with_decoder(|d| Ok(d.decode()?))
+    }
+
+    /// Decode some element according to a custom strategy. This consumes more bytes if the decoder
+    /// fails due to a lack of bytes. And error otherwise.
+    pub fn with_decoder<T>(
+        &mut self,
+        decode: impl Fn(&mut cbor::decode::Decoder<'_>) -> Result<T, Box<dyn std::error::Error>>,
+    ) -> Result<T, Box<dyn std::error::Error>> {
+        let mut should_read_more = self.bytes.is_empty();
+        let mut can_read_more = true;
+        loop {
+            if should_read_more {
+                let mut buf = [0; Self::CHUNK_SIZE];
```

Comment:
what happens if there is less than 2 MiB left for a chunk? Is there an issue when there is unused zero bytes at the end of the buffer?

Reply:
It reads `up to 2MiB`; and returns the actual size read. In case where there's no bytes left, it returns `0`. 

> Is there an issue when there is unused zero bytes at the end of the buffer

And, no. I did not alter the behavior or minicbor in this regard. If there's leftovers that the decoder do not consume, then it just returns the decoded value and ignore the surplus.